services/debugger/cache/cache.go

```diff
@@ -0,0 +1,46 @@
+package cache
+
+//go:generate mockgen -destination=../../../mocks/services/debugger/cache/mock_cache.go -package=mocks_cache github.com/rudderlabs/rudder-server/services/debugger/cache CacheAny
+
+import (
+	"github.com/rudderlabs/rudder-server/services/debugger/cache/embeddedcache"
+	"github.com/rudderlabs/rudder-server/services/debugger/cache/listcache"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+)
+
+var pkgLogger logger.Logger
+
+type CacheAny interface {
+	Cache[any]
+}
+
+type Cache[E any] interface {
+	Update(key string, value E)
+	ReadAndPopData(key string) []E
+}
+
+type Impl[E any] struct{}
+
+type FactoryImpl[E any] struct{}
+
+type Factory[E any] interface {
+	New(cacheType string) Cache[E]
+}
+
+func init() {
+	pkgLogger = logger.NewLogger().Child("cache")
+}
+
+func (FactoryImpl[E]) New(cacheType string) Cache[E] {
```

Comment:
Do we even need `FactoryImpl`? Can't we just have a `New` function to return `Cache`?

Possibly I think this whole file could be reduced to this:

```go
type cacheType int8

const (
	MemoryCacheType cacheType = iota
	BadgerCacheType
)

type Cache[T any] interface {
	Update(key string, value T)
	ReadAndPopData(key string) []T
}

func New[T any](ct cacheType, l logger.Logger) (Cache[T], error) {
	switch ct {
	case MemoryCacheType:
		l.Info("Using in-memory cache")
		return &listcache.ListCache[T]{}, nil
	case BadgerCacheType:
		l.Info("Using badger cache")
		return &embeddedcache.EmbeddedCache[T]{}, nil
	default:
		return nil, fmt.Errorf("cache type %v out of the known domain", ct)
	}
}
```

Reply:
Yeah this makes much more sense...